azure/services/agentpools/agentpools.go

```diff
@@ -77,6 +77,7 @@ func (s *Service) Reconcile(ctx context.Context) error {
 		result, err := s.CreateResource(ctx, agentPoolSpec, serviceName)
 		if err != nil {
 			resultingErr = err
+			s.scope.SetAgentPoolReady(false)
```

Comment:
Have you considered doing this at the end of the reconcile loop in the same place where we set this to true when there is no reconcile error? https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/exp/controllers/azuremanagedmachinepool_controller.go#L282 (perhaps that should also use the `s.scope.SetAgentPoolReady(true)` since that's equivalent)

The function name `SetAgentPoolReady` is maybe misleading, since it actually sets the Ready status field on the AzureManagedMachinePool CR, not the Azure agent pool resource.

Reply:
I thought about that spot, but since this particular scenario goes through the codepath ending at https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/24b93aacd3caf08d25d0bf1368e9eba4bb0e2d84/exp/controllers/azuremanagedmachinepool_controller.go#L272, it didn't seem like there was a clear way to catch where we should set ready to false.

And I'm still not sure in exactly which scenarios we'd want to set ready to false. Setting it based on whether or not this `err` is nil looks like it would include more scenarios where `status.ready` would be set to false than where I'm setting it now.
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/24b93aacd3caf08d25d0bf1368e9eba4bb0e2d84/exp/controllers/azuremanagedmachinepool_controller.go#L260

Is that closer to what we'd want anyway? Or is there some additional introspection of that `err` we'd want to do to decide what the status should be?

Reply:
> Is that closer to what we'd want anyway? Or is there some additional introspection of that err we'd want to do to decide what the status should be?

A question that might help answer this: are there any cases where you'd want the Status to still show `Ready` even though the Resource is not fully reconciled or there was an error returned from Reconcile?

Reply:
My gut reaction is to say "no," though I'm not familiar with the range of possible errors that might return. I noticed that for unmanaged AzureMachinePools, their `status.ready` is based solely on the provisioning state of the VMSS, desired replica count, and actual replica count. I can see where setting `status.ready` to false more aggressively may not necessarily be a bad thing though.

Reply:
Here is how the [Cluster API contract](https://cluster-api.sigs.k8s.io/developer/providers/machine-infrastructure.html) defines "ready" (this definition is for Machines but the same should apply to MachinePools):

`ready (boolean): indicates the provider-specific infrastructure has been provisioned and is ready`

>  for unmanaged AzureMachinePools, their status.ready is based solely on the provisioning state of the VMSS, desired replica count, and actual replica count

There might be an opportunity to standardize Ready status for AzureMachinePools and AzureManagedMachinePools.

Reply:
Ok, I think that makes sense. I took a pass at moving where we set ready to false a bit lower to right after that `svc.Reconcile` call. Let me know if that seems better. The light manual testing I've done doesn't seem to show any difference in behavior.